snapcraft/internal/sources.py

```diff
@@ -435,3 +476,54 @@ def _get_source_type_from_uri(source, ignore_errors=False):
         raise ValueError('local source is not a directory')
 
     return source_type
+
+
+def check_checksum_determine_format(source_checksum, checkfile):
+    if source_checksum.startswith('http'):
+        response = urllib.request.urlopen(source_checksum)
+        data = response.read()
+        source_checksum = data.decode('utf-8')
+        if source_checksum.index(" "):
+            source_checksum = source_checksum.split(" ", 1)[0]
+    if os.path.isfile(source_checksum):
+        filename = source_checksum
+        try:
+            filework = open(filename, 'r')
+            source_checksum = filework.read()
+            if source_checksum.index(" "):
+                source_checksum = source_checksum.split(" ", 1)[0]
+        finally:
+            filework.close()
+    check_checksum_determine_type_and_read_file(source_checksum, checkfile)
+
+
+def check_checksum_determine_type_and_read_file(source_checksum, checkfile):
+    if len(source_checksum) == 32:
+        chksum = hashlib.md5()
+    elif len(source_checksum) == 40:
+        chksum = hashlib.sha1()
+    elif len(source_checksum) == 56:
+        chksum = hashlib.sha224()
+    elif len(source_checksum) == 64:
+        chksum = hashlib.sha256()
+    elif len(source_checksum) == 96:
+        chksum = hashlib.sha384()
+    elif len(source_checksum) == 128:
+        chksum = hashlib.sha512()
+    else:
+        raise IncompatibleOptionsError("Invalid checksum format")
```

Comment:
This could be a dictionary instead

```python
_HASH_FUNCTIONS = {
    32: hashlib.md5,
    40: hashlib.sha1,
    ...
    ...
}

try:
     checksum_function = _HASH_FUNCTIONS[len(source_checksum)]()
except KeyError:
    raise IncompatibleOptionsError("Invalid checksum format")